### ISSUE TITLE
100% usage, freeze, when rotating preview toward horizon

### ISSUE THREAD

**Issue by jason-green-io** [llbit/chunky#320]
Linux theman 4.0.4-2-ARCH #1 SMP PREEMPT Fri May 22 03:05:23 UTC 2015 x86_64 GNU/Linux

java version "1.8.0_45"
Java(TM) SE Runtime Environment (build 1.8.0_45-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.45-b02, mixed 

java -jar chunky.jar

I give it 2048M of ram, need to know anything else?

**Comment by modwizcode** [comment]
I've seen the same issue, I used a profiler to check where the cpu was spending its time and found that the code was stuck in awt doing `drawLine` or some such thing, I plan to investigate further, I haven't encountered this on windows, may be a subtle bug in chunky's behavior under linux.

**Comment by BlindSigma** [comment]
I've experienced this as well under Linux; I haven't been able to do any new renders lately as a result. I don't think it has been resolved.

**Comment by modwizcode** [comment]
Found the problem, it's an issue with the code that finds the intersection of the camera boundaries with the chunks. This is used to power the overview map's yellow indicator of camera viewport. Sometimes, due to very close to parallel, but not quite, rays, the distance found can be thousands and thousands of pixels. When the code goes to draw the line, it calls into the Java awt code to draw a line, however due to a bug in the Java implementation of graphics, possibly related to linux, this call just goes forever, doing who knows what.

I fixed this for myself by checking if the y direction of the cast ray was within `.0001` of zero, then set the component to zero (yielding a null boundary), not sure why this doesn't cause odd behavior, but I'll work on a proper patch and contribute it later.

**Comment by modwizcode** [comment]
After more investigation into this problem, I've found that while the JVM implementation on Linux is the cause of the 100% usage crash (and should be reported to Oracle), the issue with the wildly out of bounds coordinates is likely a glitch in the math used by Chunky to render the view cone of the camera. Patching the drawLine calls to cap at the boundaries fixes the issue but properly reviews the oddly angled lines. 

I'll PR the fix in to hotfix this issue and work towards a proper solution.

**Comment by DaMatrix** [comment]
Can still confirm this issue. Sorry to bump, but this should be fixed soon. I haven't made any renders for a while because of this.

**Comment by llbit** [comment]
I have not really been following this issue but it looks like this should be high priority to fix.

Thanks for the info MoeHouse. Not sure why that would cause a hang on Linux, but I'll do some testing and see if I can reproduce it.

**Comment by llbit** [comment]
I spent some time trying to reproduce the issue in an Ubuntu VM, but couldn't get it to lock up. Are you able to reproduce the issue with a completely empty scene, or does it only happen if you have loaded any chunks?

Is it possible to extract the camera position and orientation information (Camera tab) to reproduce the issue?

**Comment by DaMatrix** [comment]
I can send you the setup I've been using, but it locks up after a seemingly random amount of time, and (at least for me) only while the render is running.

**Comment by llbit** [comment]
@DaMatrix Yes, please send me anything to help reproduce the issue. Can you tell me the exact steps I should take to reproduce it?

**Comment by DaMatrix** [comment]
I'll send my configuration, what happens (for me) is it renders anywhere between 0-400 samples and then just completely locks up.  Can't send it now unfortunately, I don't have access to my computer.

**Comment by llbit** [comment]
I was not able to reproduce this issue even though I tried with Java 1.8.0_45 which Jason reported having issues with. However, I rewrote the code for the camera field of view indicator to clip the drawn lines to the visible part of the 2D map. This should prevent an inefficient drawLine implementation in Java from crashing Chunky. If that was the cause of the issue, then it should be fixed now.

Since I can't reproduce the bug myself I will wait and see if the issue remains after releasing the fix.
